libbeat/processors/actions/uppercase_fields.go

```diff
@@ -0,0 +1,126 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	"github.com/elastic/beats/v7/libbeat/processors/checks"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/pkg/errors"
+)
+
+type upperCaseProcessor struct {
+	Fields        []string
+	IgnoreMissing bool
+	FailOnError   bool
+}
+
+func init() {
+	processors.RegisterPlugin(
+		"uppercase_fields",
+		checks.ConfigChecked(
+			NewUpperCaseProcessor,
+			checks.RequireFields("fields"),
+			checks.AllowedFields("fields", "when", "ignore_missing", "fail_on_error"),
+		),
+	)
+}
+
+func NewUpperCaseProcessor(c *conf.C) (processors.Processor, error) {
+	config := struct {
+		Fields        []string `config:"fields"`
+		IgnoreMissing bool     `config:"ignore_missing"`
+		FailOnError   bool     `config:"fail_on_error"`
+	}{
+		IgnoreMissing: false,
+		FailOnError:   true,
+	}
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("failed to unpack the uppercase_fields configuration: %s", err)
+	}
+
+	// Skip mandatory fields
+	for _, readOnly := range processors.MandatoryExportedFields {
+		for i, field := range config.Fields {
+			if field == readOnly {
+				config.Fields = append(config.Fields[:i], config.Fields[i+1:]...)
+			}
+		}
+	}
+
+	return &upperCaseProcessor{Fields: config.Fields, IgnoreMissing: config.IgnoreMissing, FailOnError: config.FailOnError}, nil
+}
+
+func (p *upperCaseProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	var backup *beat.Event
+	if p.FailOnError {
+		backup = event.Clone()
+	}
+
+	for _, field := range p.Fields {
+		if err := p.upperCaseField(event, field); err != nil {
```

Comment:
There is a lot of code duplication in these two processors. The only part that differs is what function we're applying to the field name (last segment of the path).

I think it would be better to implement a `changeFieldProcessor` and a parameter called `changeFunc`. We would avoid a lot of code duplication if we did this.

It will be still exposed as `upperCaseFieldProcessor` and `lowerCaseFieldProcessor` for external use but in their constructor function they would just create and return a `changeFieldProcessor` with `toLower` or `toUpper` function as a parameter.

Reply:
Just to clarify, since this comment and your first comment are kind of confusing to me.

You mentioned on your first comment that these 2 processors can be replaced by one so why do I need to keep their *CaseFieldProcessor struct?

Also, where should I put the changeFieldProcessor? should I create it in another file? 
Or should I create a new file change_fields.go that will include the changeFieldProcessor struct with both of the lowerCase/upperCase constructors that will return a changeFieldProcessor each with their relevant changeFunc?